presto-hive/src/main/java/com/facebook/presto/hive/statistics/QuickStatsProvider.java

```diff
@@ -0,0 +1,318 @@
+    }
+
+    public Map<String, PartitionStatistics> getQuickStats(ConnectorSession session, SemiTransactionalHiveMetastore metastore, SchemaTableName table,
+            MetastoreContext metastoreContext, List<String> partitionIds)
+    {
+        if (!isQuickStatsEnabled(session)) {
+            return partitionIds.stream().collect(toMap(k -> k, v -> empty()));
+        }
+
+        CompletableFuture<PartitionStatistics>[] partitionQuickStatCompletableFutures = new CompletableFuture[partitionIds.size()];
+        for (int counter = 0; counter < partitionIds.size(); counter++) {
+            String partitionId = partitionIds.get(counter);
+            partitionQuickStatCompletableFutures[counter] = supplyAsync(() -> getQuickStats(session, metastore, table, metastoreContext, partitionId), backgroundFetchExecutor);
+        }
+
+        try {
+            // Wait for all the partitions to get their quick stats
+            // If this query is reading a partition for which we do not already have cached quick stats,
+            // we will block the execution of the query until the stats are fetched for all such partitions,
+            // or we time out waiting for the fetch
+            allOf(partitionQuickStatCompletableFutures).get(getQuickStatsBuildTimeoutMillis(session), MILLISECONDS);
+        }
+        catch (InterruptedException | ExecutionException e) {
+            log.error(e);
+            throw new RuntimeException(e);
+        }
+        catch (TimeoutException e) {
+            log.warn(e, "Timeout while building quick stats");
+            session.getRuntimeStats().addMetricValue("QuickStatsProvider/QuickStatsBuildTimeout", RuntimeUnit.NONE, 1L);
+        }
+
+        ImmutableMap.Builder<String, PartitionStatistics> result = ImmutableMap.builder();
+        for (int counter = 0; counter < partitionQuickStatCompletableFutures.length; counter++) {
+            String partitionId = partitionIds.get(counter);
+            CompletableFuture<PartitionStatistics> future = partitionQuickStatCompletableFutures[counter];
+            if (future.isDone() && !future.isCancelled() && !future.isCompletedExceptionally()) {
+                try {
+                    result.put(partitionId, future.get());
+                }
+                catch (InterruptedException | ExecutionException e) {
+                    // This should not happen because we checked that the future was completed successfully
+                    log.error(e, "Failed to get value for a quick stats future which was completed successfully");
+                    throw new RuntimeException(e);
+                }
+            }
+            else {
+                // If a future did not finish, or finished exceptionally, we do not add it to the results
+                // A new query for the same partition could trigger a successful quick stats fetch for this partition
+                result.put(partitionId, empty());
```

Comment:
This would be identical to current behavior where a query comes in for say 10 partitions, and only 5 of them have HMS provided stats. See [MetastoreHiveStatisticsProvider#getTableStatistics](https://github.com/prestodb/presto/blob/496e98ca25fb243ab25731d188c2fd56eba3b3e9/presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java#L389) for how the rollup happens. Briefly, it aggregates the available stats and comes up with an underestimate